src/support/generation.c

```diff
@@ -98,119 +98,141 @@ __wt_gen_next_drain(WT_SESSION_IMPL *session, int which)
 }
 
 /*
- * __wt_gen_drain --
- *     Wait for the resource to drain.
+ * __gen_drain_callback --
+ *     Wait for single session's generation to drain, callback from the session array walk.
  */
-void
-__wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
+static void
+__gen_drain_callback(WT_SESSION_IMPL *session, bool *exit_walkp, void *cookiep)
 {
-    struct timespec start, stop;
+    struct timespec stop;
     WT_CONNECTION_IMPL *conn;
-    WT_SESSION_IMPL *s;
-    uint64_t minutes, time_diff_ms, v;
-    uint32_t i, session_cnt;
-    int pause_cnt;
-    bool verbose_timeout_flags;
+    WT_GENERATION_DRAIN_COOKIE *cookie;
+    uint64_t time_diff_ms, v;
 
+    cookie = (WT_GENERATION_DRAIN_COOKIE *)cookiep;
     conn = S2C(session);
-#ifdef HAVE_DIAGNOSTIC
-    verbose_timeout_flags = false;
-#else
-    WT_UNUSED(verbose_timeout_flags);
-#endif
-    __wt_epoch(NULL, &start);
 
-    /*
-     * No lock is required because the session array is fixed size, but it may contain inactive
-     * entries. We must review any active session, so insert a read barrier after reading the active
-     * session count. That way, no matter what sessions come or go, we'll check the slots for all of
-     * the sessions that could have been active when we started our check.
-     */
-    WT_ORDERED_READ(session_cnt, conn->session_cnt);
-    for (minutes = 0, pause_cnt = 0, s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
-        if (!s->active)
-            continue;
+    for (;;) {
+        /* Ensure we only read the value once. */
```

Comment:
I see Andrew's left a comment and it's been resolved - what was the outcome there? The comment and the code disagree.

Reply:
I'm just clarifying as I'm not sure which of Andrew's comments you are referring to. The comment of "Ensure we read the value once." is incorrect but persisted from the previous implementation. We won't convert it to WT_READ_ONCE in this Pr.

Reply:
Oh right, now I see it. Yes we won't change it, it was just a note.

Reply:
OK, thanks. Is it reasonable to add a `FIXME-WT-XXXX` then? Or are we tracking these `READ_ONCE` candidates elsewhere.

Reply:
See #9702

Reply:
Oh, nice - looks like that's tracking it. Thanks